python/pyspark/sql/datasource.py

```diff
@@ -0,0 +1,184 @@
+
+    @property
+    def name(self) -> str:
+        """
+        Returns a string represents the short name of this data source.
+        """
+        return self.__class__.__name__
+
+    def schema(self) -> Union[StructType, str]:
+        """
+        Returns the schema of the data source.
+
+        It can reference the `options` field to infer the data source's schema when
+        users do not explicitly specify it. This method is invoked once when calling
+        `spark.read.format(...).load()` to get the schema for a data source read operation.
+        If this method is not implemented, and a user does not provide a schema when
+        reading the data source, an exception will be thrown.
+
+        Returns
+        -------
+        schema : StructType or str
+            The schema of this data source or a DDL string represents the schema
+
+        Examples
+        --------
+        Returns a DDL string:
+
+        >>> def schema(self):
+        >>>    return "a INT, b STRING"
+
+        Returns a StructType:
+
+        >>> def schema(self):
+        >>>   return StructType().add("a", "int").add("b", "string")
+        """
+        ...
+
+    def reader(self, schema: StructType) -> "DataSourceReader":
+        """
+        Returns a DataSourceReader instance for reading data.
+
+        This method is required for readable data sources. It will be called once during
+        the physical planning stage in the Spark planner.
+
+        Parameters
+        ----------
+        schema : StructType
+            The schema of the data to be read.
+
+        Returns
+        -------
+        reader : DataSourceReader
+            A reader instance for this data source.
+        """
+        raise NotImplementedError
+
+
+class DataSourceReader(ABC):
+    """
+    A base class for data source readers. Data source readers are responsible for
+    outputting data from a data source.
+    """
+
+    def partitions(self) -> Iterator[Any]:
+        """
+        Returns a list of partitions for this data source.
+
+        This method is called once during the physical planning stage to generate a list
+        of partitions. If the method returns N partitions, then the planner will create
+        N tasks. Each task will then execute `read(partition)` in parallel, using each
+        partition value to read the data from this data source.
+
+        If this method is not implemented, or returns an empty list, Spark will create one
+        partition for the result DataFrame and use one single task to read the data.
+
+        Returns
+        -------
+        partitions : list
+            A list of partitions for this data source. The partition can be any arbitrary
+            serializable objects.
+
+        Notes
+        -----
+        This method should not return any un-picklable objects.
+
+        Examples
+        --------
+        Returns a list of integers:
+
+        >>> def partitions(self):
+        >>>     return [1, 2, 3]
```

Comment:
```suggestion
        ...     return [1, 2, 3]
```

and should update below too

Reply:
Good catch!